### PR TITLE
Integration file methods

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Add methods to support file management during integration experiments.
- Modify archive cache use to make symbolic links rather than copying archives.

### CODE REVIEW

File: integration/apps/apps.py [hunk 1]
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
...
+                                 the trial was be executed.  The value
+                                 of this parameter will be the same
+                                 for all calls made to this method
+                                 during an experiment.
+
+        '''
+        pass
+
     def get_bash_setup_commands(self):
+
```
Comment (cmcantalupo) [inline integration/apps/apps.py]: Yuck

File: integration/apps/apps.py [hunk 2]
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
         '''
         return None
 
+
+    def prelaunch_setup(self, run_id, output_dir):
+        '''Execute any commands that will create files to be used during a
+           launch of an step in an experiment.  This may include
```
Comment (dianarg) [inline integration/apps/apps.py]: a step

File: integration/apps/apps.py [hunk 3]
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
...
+           creating symbolic links, copying files into the output
+           directory, or executing commands that create input data for
+           the application.  This step will be called repeately, once
```
Comment (dianarg) [inline integration/apps/apps.py]: repeatedly

File: integration/apps/apps.py [hunk 4]
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
...
+                                 same for all calls made to this
+                                 method during an experiment.
+
+        '''
+        pass
+
+    def postlaunch_teardown(self, run_id, output_dir):
+        '''Delete files or symbolic links created as part of an experiment
+           trial that are not needed by subsequent trials.  This step
+           will be called repeately, once for each trial executed by
```
Comment (dianarg) [inline integration/apps/apps.py]: repeatedly

File: integration/apps/apps.py [hunk 5]
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
...
+               output_dir (str): Path to the working directory where
+                                 the trial was be executed.  The value
+                                 of this parameter will be the same
+                                 for all calls made to this method
+                                 during an experiment.
+
+        '''
+        pass
+
+    def experiment_teardown(self, output_dir):
```
Comment (dianarg) [inline integration/apps/apps.py:189]: Can we have an experiment_setup() too?

**Review by dianarg — CHANGES_REQUESTED** [review]

File: integration/experiment/launch_util.py [hunk 6]
```diff
@@ -141,13 +141,16 @@ def launch_all_runs(targets, num_nodes, iterations, extra_cli_args, output_dir,
...
+            app_conf.prelaunch_setup(run_id, output_dir)
             launch_run(agent_conf, app_conf, run_id, output_dir,
                        extra_cli_args=extra_cli_args,
                        num_nodes=num_nodes, enable_traces=enable_traces,
                        enable_profile_traces=enable_profile_traces)
+            app_conf.postlaunch_teardown(run_id, output_dir)
 
             # rest to cool off between runs
             time.sleep(cool_off_time)
+    app_conf.experiment_teardown(output_dir)
```
Comment (dianarg) [inline integration/experiment/launch_util.py]: this only happens for the last app_conf; loop through all of them?

File: integration/apps/apps.py [hunk 7]
```diff
@@ -137,6 +137,79 @@ def get_cpu_per_rank(self):
...
+
+    def experiment_setup(self, output_dir):
+        '''Create files to be used throughout an experiment and present for
+           all trials.  These files typically persist unchanged
+           throughout the execution of the experiement and are cleaned
+           up with the experiement_teardown() method.  This method may
+           create symbolic links, copy files into the output
+           directory, or execute commands that create input data for
+           the application.
+
```
Comment (dianarg) [inline integration/apps/apps.py]: docstring for output_dir

File: integration/apps/apps.py [hunk 8]
```diff
@@ -137,6 +137,79 @@ def get_cpu_per_rank(self):
...
+                                 for all calls made to this method
+                                 during an experiment.
+
+        '''
+        pass
+
+    def experiment_setup(self, output_dir):
+        '''Create files to be used throughout an experiment and present for
+           all trials.  These files typically persist unchanged
+           throughout the execution of the experiement and are cleaned
```
Comment (dianarg) [inline integration/apps/apps.py]: experiment, more below

File: integration/apps/nekbone/nekbone.py [hunk 9]
```diff
@@ -54,25 +54,32 @@ def __init__(self, add_barriers=False):
...
 
     def get_rank_per_node(self):
         # TODO: use self._machine_file to determine
         return self._num_rank_per_node
 
-    def get_bash_setup_commands(self):
+    def trial_setup(self, run_id, output_dir):
         size = 10000  # this size varies per system
-        input_file = textwrap.dedent('''
+        input_data = textwrap.dedent('''
```
Comment (bgeltz) [inline integration/apps/nekbone/nekbone.py:65]: The data.rea should always be the same for every trial.

Doesn't this belong in experiment_setup?

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: integration/apps/nekbone/nekbone.py (hunk 9 above)
Comment (cmcantalupo) [reply integration/apps/nekbone/nekbone.py:65]: We got rid of experiment_setup.  The issue is that there are N app_conf objects, but only one start to the experiment.  Instead, the implementer of the trial_setup() is responsible for checking for file existence.

**Review by dianarg — APPROVED** [review]

**Review by bgeltz — APPROVED** [review]

**Pull request merged**